notebooks/_06-Architecture.md

```diff
@@ -0,0 +1,429 @@
+  
+  - If using **personal laptop**: 
+      * Create a conda virtual environment with all the necessary packages for the project, and push it to the github repository as a .yml file in order for it to be duplicated by other members and end users, **OR**
+      * Create a Docker image of the computational environment to be used by all team members.
+
+### 3.3  How will one test that the computational environment is correctly created?
+
+- Shell commands in either situation will be used to check that the correct environment was created/installed.
+- Write class/function to make sure each of the needed packages and right python version is installed.
+
+## 4. Description of major classes to be used (relates to major components above):
+<!--- **Class responsibilities**
+- **Interactions between classes**
+- **Class hierarchies / state transitions / object persistence**
+- **Organization of classes into subsystems (if necessary / planned)**
+- **Why does each class have “jurisdiction” over the parts of the system that it does?** -->
+
+The classes are to be used in the simulation work are as follows:
+ -  Class that contains functions to simulate data based on best fit distributions
+ -  Class that contains functions to simulate data based on assumed causal graph
+ -  Class to estimate the choice model based on specified utility equations and retrieve model statistics
+
+## 5. Description of the minimally viable product (if any):
+- **Note the portions of the architecture responsible for producing the minimally viable product.**
+
+ - A minimally viable product is a notebook that takes data from outside sources and imports scripts including classes
+ from question 4, calls on the functions in each of the classes to simulate data, estimate models, and provide
+ parameters and statistics relevant to the problem at hand.
+
+## 6. Description of algorithms:
+### 6.1  What, if any, computational algorithms are being implemented as part of this project instead of relying upon external implementations?
+Hassan to fill in based on the deconfounder work.
+
+### 6.2 What purpose(s) do those algorithms serve?
+See above
+
+### 6.3 Are each of these algorithms fully understood by someone on the project team?
+See above
+
+### 6.4  What are alternative algorithms that could have served the purposes of the algorithms we are implementing?
+See above
+
+### 6.5 Why are we using the particular algorithms we’ve chosen instead of others? 
+See above
+
+### 6.6 Why do we believe the algorithms can be implemented?
+See above
+
+### 6.7 Do we have any concerns with regard to algorithmic efficiency?
+See above
+
+
+## 7. Description of data usage in the system:
+### 7.1 How will data be ingested / accessed?
+Data will be ingested using pandas from either an online path to the dataset or from a local path on the user's machine
+
+### 7.2 How will data be validated?
+Data can be validated using third party dependencies such
```

Comment:
If I'm not mistaken, one's estimation class will essentially have two sources of data:
- Simulations based on fitted distributions to actual data
- Simulations based on pre-specified distributions for all nodes in one's causal graph.

I imagine there should be some simple validation to make sure that one gets the expected data back from the simulation. For instance, do all the simulated columns have the correct data types and ranges for numeric values? Are the means and variances of the continuous variables similar to what one expect?